kong/runloop/plugins_iterator.lua

```diff
@@ -211,22 +253,85 @@ local plugins_iterator_mt = { __call = get_next }
 --
 -- @param[type=table] ctx Nginx context table
 -- @param[type=string] phase Plugins iterator execution phase
--- @param[type=table] plugins Plugins table
 -- @treturn function iterator
-local function plugins_iterator(ctx, phase, plugins_plan)
+local function iterate(self, ctx, phase)
   if not ctx.plugins then
     ctx.plugins = {}
   end
 
-  local plugins_iterator_state = {
-    i = 0,
-    ctx = ctx,
+  local iteration = {
+    iterator = self,
     phase = phase,
-    plugins_plan = plugins_plan,
+    ctx = ctx,
+    i = 0,
   }
 
-  return setmetatable(plugins_iterator_state, plugins_iterator_mt)
+  return get_next, iteration
+end
+
+
+function PluginsIterator.new(version)
+  loaded_plugins = loaded_plugins or get_loaded_plugins()
+
+  local map = {}
+  local combos = {}
+  local phases
+  if subsystem == "stream" then
+    phases = {
+      init_worker = {},
+      preread     = {},
+      log         = {},
+    }
+  else
+    phases = {
+      init_worker   = {},
+      certificate   = {},
+      rewrite       = {},
+      access        = {},
+      header_filter = {},
+      body_filter   = {},
+      log           = {},
+    }
+  end
+
+  for plugin, err in kong.db.plugins:each(1000) do
+    if err then
+      return nil, err
+    end
+
+    if should_process_plugin(plugin) then
+      map[plugin.name] = true
+
+      local combo_key = (plugin.route    and 1 or 0)
+                      + (plugin.service  and 2 or 0)
+                      + (plugin.consumer and 4 or 0)
+
+      combos[plugin.name] = combos[plugin.name] or {}
+      combos[plugin.name][combo_key] = true
+    end
+  end
+
+  local phase_handler
+  for _, plugin in ipairs(loaded_plugins) do
+    if combos[plugin.name] then
+      for phase_name, phase in pairs(phases) do
+        phase_handler = plugin.handler[phase_name]
+        if phase_handler and phase_handler ~= BasePlugin[phase_name] then
```

Comment:
Should this check for `"function"`, like here:
https://github.com/Kong/kong/pull/4590/files#diff-1f0f137350308286de76ec59740c8edaR412

Reply:
Sure, I will make the change